File: app/src/main/java/d/d/meshenger/Contacts.java
package d.d.meshenger;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


public class Contacts {
    private List<Contact> contacts;

    public Contacts() {
        contacts = new ArrayList<>();
    }

    public List<Contact> getContactList() {
        return contacts;
    }

    public List<Contact> getContactListCopy() {
        return new ArrayList<>(contacts);
    }

    public void addContact(Contact contact) {
        int idx = findContact(contact.getPublicKey());
        if (idx >= 0) {
            // contact exists - replace
            contacts.set(idx, contact);
        } else {
            contacts.add(contact);
        }
    }

    public void deleteContact(byte[] publicKey) {
        int idx = findContact(publicKey);
        if (idx >= 0) {
            this.contacts.remove(idx);
        }
    }

    private int findContact(byte[] publicKey) {
        for (int i = 0; i < contacts.size(); i += 1) {
            if (Arrays.equals(contacts.get(i).getPublicKey(), publicKey)) {
                return i;
            }
        }
        return -1;
    }

    public Contact getContactByPublicKey(byte[] publicKey) {
        for (Contact contact : contacts) {
            if (Arrays.equals(contact.getPublicKey(), publicKey)) {
                return contact;
            }
        }
        return null;
    }

    Contact getContactByName(String name) {
        for (Contact contact : contacts) {
            if (contact.getName().equals(name)) {
                return contact;
            }
        }
        return null;
    }

    public static JSONObject toJSON(Contacts contacts) throws JSONException {
        JSONObject obj = new JSONObject();

        JSONArray array = new JSONArray();
        for (Contact contact : contacts.contacts) {
            array.put(Contact.toJSON(contact));
        }
        obj.put("entries", array);

        return obj;
    }

    public static Contacts fromJSON(JSONObject obj) throws JSONException {
        Contacts contacts = new Contacts();

        JSONArray array = obj.getJSONArray("entries");
        for (int i = 0; i < array.length(); i += 1) {
            contacts.contacts.add(
                Contact.fromJSON(array.getJSONObject(i))
            );
        }
        return contacts;
    }
}

File: app/src/main/kotlin/d/d/meshenger/Contacts.kt
package d.d.meshenger

import d.d.meshenger.mock.MockContacts
import org.json.JSONArray

import org.json.JSONException

import org.json.JSONObject
import java.util.*
import kotlin.collections.ArrayList

//TODO: Mock Objects are active here
class Contacts {

    private val contacts: ArrayList<Contact?> = MockContacts.generateMockContactList()
//    ArrayList()

    companion object {
        @Throws(JSONException::class)
        fun toJSON(contacts: Contacts): JSONObject {
            val obj = JSONObject()
            val array = JSONArray()
            for (contact in contacts.contacts) {
                array.put(Contact.toJSON(contact!!))
            }
            obj.put("entries", array)
            return obj
        }

        @Throws(JSONException::class)
        fun fromJSON(obj: JSONObject): Contacts {
            val contacts = Contacts()
            val array = obj.getJSONArray("entries")
            var i = 0
            while (i < array.length()) {
                contacts.contacts.add(
                    Contact.fromJSON(array.getJSONObject(i))
                )
                i += 1
            }
            return contacts
        }
    }

    fun getContactList(): ArrayList<Contact?> = contacts

    fun getContactListCopy(): ArrayList<Contact?> = ArrayList(contacts)

    fun addContact(contact: Contact) {
        val idx = findContact(contact.publicKey)
        if (idx >= 0) {
            // contact exists - replace
            contacts[idx] = contact
        } else {
            contacts.add(contact)
        }
    }

    fun deleteContact(publicKey: ByteArray?) {
        val idx = findContact(publicKey)
        if (idx >= 0) {
            contacts.removeAt(idx)
        }
    }

    private fun findContact(publicKey: ByteArray?): Int {
        var i = 0
        while (i < contacts.size) {
            if (Arrays.equals(contacts[i]!!.publicKey, publicKey)) {
                return i
            }
            i += 1
        }
        return -1
    }

    fun getContactByPublicKey(publicKey: ByteArray?): Contact? {
        for (contact in contacts) {
            if (Arrays.equals(contact!!.publicKey, publicKey)) {
                return contact
            }
        }
        return null
    }

    fun getContactByName(name: String): Contact? {
        for (contact in contacts) {
            if (contact!!.name == name) {
                return contact
            }
        }
        return null
    }



}